### PR TITLE
Sprint1

### PR DESCRIPTION
**Pull request by larabr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,634 +1 @@
...
-    var ObjectId = mongoose.Types.ObjectId;
-    
-    var artist = new Artist({
-      _id : ObjectId()
-    });
-
-    var Track = new Track({
-      artist: artist._id
-    });
+## JobAdvisor
```
Comment (aronfiechter) [inline README.md:1]: Maybe we could add more to the README? Don't really know what though for now.

File: models/Review.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+'use strict';
+const mongoose = require('mongoose');
+const ObjectID = mongoose.Schema.Types.ObjectId;
+require ('./Freelance');
+
+const Review = exports.Review = new mongoose.Schema({
+	//in future maybe add User and author is ObjectID ref User
+		author 	: { type: String, required: true },
+		title 	: { type: String },
+		text 	: { type: String },
```
Comment (aronfiechter) [inline models/Review.js]: On GitHub indentation is sometimes weirded out, but this line and line 24 just afterwards are actually wrongly indented in the actual file.

File: models/Review.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+);
+
+Review.pre('save', function(next) {
+	if (this.score > 5){
+		this.score = 5;
+	} else if (this.score < 0) {
+		this.score = 0;
+	}
+
+	if (this.text !== undefined) {
```
Comment (aronfiechter) [inline models/Review.js:38]: I think you can just check `if (this.text)`

File: models/Review.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		let length = this.text.length;
+
+		if(length > 1000) {
```
Comment (aronfiechter) [inline models/Review.js]: Space after `if`

File: models/Tag.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+'use strict';
+const mongoose = require('mongoose');
+const ObjectID = mongoose.Schema.Types.ObjectId;
+require ('./Freelance');
+
+const Tag = exports.Tag = new mongoose.Schema({
+	//in future maybe add User and author is ObjectID ref User
+		tagName 		: { type: String, required: true },
+		freelancers 	: [{ type: ObjectID, ref: 'Freelance', default:[] }],
```
Comment (aronfiechter) [inline models/Tag.js]: One extra indent before the `:`

File: package.json [hunk 6]
```diff
@@ -1,42 +1,37 @@
...
+  "description": "JobAdvisor - Software Engineering Atelier 2017",
   "main": "app.js",
   "scripts": {
     "install": "chmod -R u+x ./bin/ && ./bin/installSeleniumAndChromeWebDriver",
-    "start": "DEBUG='atelierbeats-server' ./bin/www",
-    "test": "npm run test-mocha && npm start | node seed.js && ./nightwatch",
-    "test-mocha": "JUNIT_REPORT_PATH=./jenkins-test-results.xml istanbul cover --report cobertura _mocha -- --timeout 3000 -R spec ./test/ex1 ./test/ex2 ./test/ex3 ./test/ex4 --colors --reporter mocha-jenkins-reporter",
-    "test-nightwatch": "node seed.js && ./nightwatch"
+    "start": "DEBUG='jobadvisor' ./bin/www",
+    "start:dev": "DEBUG='jobadvisor' nodemon ./bin/www -e html,css,dust",
```
Comment (aronfiechter) [inline package.json]: Add `,js,json,sh` for option `-e`

File: public/components/freelancer.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  const tags = data.tags;
+  const listTags = tags.map((tag, index) =>
+    <li key={index}>
+      {tag.tagName}
+    </li>
+  );
+
+  ReactDOM.render(
+    <FreelancerView
+      first={data.firstName}
```
Comment (aronfiechter) [inline public/components/freelancer.js:21]: For all the following lines, does spacing around `=` work or is there none because it's HTML?

File: public/components/homepage.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  render () {
+    return (<button id="freelancer-create-btn" onClick={this.redirectFreelancerForm}>Add a new freelancer</button>);
+  }
+}
+
+// Card which displays a freelancer's information
+class FreelancerCard extends React.Component {
+  redirectFreelancer(freelancer) {
+    return function() {
+      document.location = `/freelance/${freelancer.props._id}`;
```
Comment (aronfiechter) [inline public/components/homepage.js:94]: There's a property array `links` in each freelancer that contains the links. The link to the freelancer itself (`"rel" : "self"`) can be found at `freelancer.links[0].href`, and it's the full link to the freelancer.
If we use that, if we change the routing then the change will be automatic for the client side.

File: public/css/style.css [hunk 9]
```diff
@@ -1,57 +1,21 @@
-body {
-  padding: 50px;
-  font: 16px "Lucida Grande", Helvetica, Arial, sans-serif;
+* {
```
Comment (aronfiechter) [inline public/css/style.css:1]: Would it be possible to include this file in all other `CSS` files, and save all colours in this general style file using CSS variables? (Like some of us did in DiscardHumanity).

File: public/js/ajax.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ * @param  {Object}   data     body data
+ * @param  {Function} callback on response
+ */
+
+function ajaxRequest(method, url, headers, data, callback) {
+	let request = new XMLHttpRequest();
+	request.open(method, url, true);
+	request.onreadystatechange = function() {
+		if (request.readyState === 4) {
+			if (request.status === 200) {
```
Comment (aronfiechter) [inline public/js/ajax.js]: Would a `switch` statement work here?

File: routes/freelance/router.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    });
+  } else res.sendStatus(400);
+});
+
+// GET freelance/:freelanceid
+router.get('/:freelanceid', function(req, res, next) {
+
+  if (ObjectId.isValid(req.params.freelanceid)) {
+    // distinguish between raw and ajax GET request (to render page or return JSON)
+    if(req.headers.ajax) {
```
Comment (aronfiechter) [inline routes/freelance/router.js:32]: As said, we should consider splitting this into two routes `/freelance/:id` and `/freelance/:id/view`

File: routes/search/router.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+const mongoose = require('mongoose');
+const ObjectId = mongoose.Types.ObjectId;
+const Freelance = mongoose.model('Freelance');
+
+// Supported methods
+router.all('/', middleware.supportedMethods('GET, OPTIONS'));
+
+// Return index
+router.get('/', function(req, res, next) {
+  const keyword = req.query.keyword;
```
Comment (aronfiechter) [inline routes/search/router.js:18]: Later, we should split `keyword` on `" "` and create a filter for each word in the keyword.

File: seed_data/seedData.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+var reviews = require('./reviewsData');
+var reviewsSize = reviews.data.length;
+var tags = require('./tagsData');
+var tagsSize = tags.data.length;
+var categories = require('./categoriesData');
+var catSize = categories.data.length;
+
+//set the category
+for (var i = freelancers.data.length - 1; i >= 0; i--) {
+  var n = Math.floor(Math.random()*catSize);
```
Comment (aronfiechter) [inline seed_data/seedData.js]: Spaces around `*`

File: seed_data/seedData.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  // console.log(categories.data[n]['_id']);
+  var rndId = categories.data[n]['_id'];
+  freelancers.data[i]['category'] = rndId;
+  categories.data[n].freelancers.push(freelancers.data[i]['_id']);
+}
+// bind reviews
+for (var i = reviewsSize - 1; i >= 0; i--) {
+  var id = reviews.data[i]['_id'];
+  var n = Math.floor(Math.random()*freelancersSize);
```
Comment (aronfiechter) [inline seed_data/seedData.js]: Spaces around `*`

File: seed_data/seedData.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  freelancers.data[n].reviews.push(id);
+}
+// bind tags
+for (var i = freelancersSize - 1; i >= 0; i--) {
+  var m = Math.floor(Math.random()*5 + 1); // give each freelancrs 1-5 random tags
```
Comment (aronfiechter) [inline seed_data/seedData.js]: Spaces around `*`

File: seed_data/seedData.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  for (var j = 0; j < m; j++) {
+    var n = Math.floor(Math.random()*tagsSize);
```
Comment (aronfiechter) [inline seed_data/seedData.js]: Spaces around `*`

File: test/models/constructAndCreate/Freelance.js [hunk 17]
```diff
@@ -0,0 +1,172 @@
+/* tests for constructor and creation of Freelance */
```
Comment (aronfiechter) [inline test/models/constructAndCreate/Freelance.js:1]: On the whole file: spaces before `{` and after `}` in some cases

File: test/models/constructAndCreate/Review.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
+/* tests for constructor and creation of Review */
```
Comment (aronfiechter) [inline test/models/constructAndCreate/Review.js:1]: Same as for `Freelance.js`

File: test/models/utils.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+/* Utilities for the models tests */
```
Comment (aronfiechter) [inline test/models/utils.js:1]: Same as for `Freelance.js`

File: test/routes-test/category.js [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+const app = require(config.projectRoot + '/app');
+const seedData = require('../../seed_data/seedData');
+const seedDb = require('../../seed_data/seedDb');
+const utils = require('../../seed_data/utils');
+const routersutils = require('./utils');
+const request = require('supertest');
+
+describe('Category test: ', function() {
+
+  describe('GET /category/', function(){
```
Comment (aronfiechter) [inline test/routes-test/category.js]: Space before `{`

File: test/routes-test/freelance-get.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+var app = require(config.projectRoot + '/app');
+var seedData = require('../../seed_data/seedData');
+var seedDb = require('../../seed_data/seedDb');
+var utils = require('../../seed_data/utils');
+var freelanceutils = require('./utils');
+var request = require('supertest');
+
+describe('Freelance-get test: ', function() {
+
+  describe('GET /freelance/:freelanceid', function(){
```
Comment (aronfiechter) [inline test/routes-test/freelance-get.js]: Space before `{`

File: test/routes-test/root.js [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+var app = require(config.projectRoot + '/app');
+var seedData = require('../../seed_data/seedData');
+var seedDb = require('../../seed_data/seedDb');
+var utils = require('../../seed_data/utils');
+var freelanceutils = require('./utils');
+var request = require('supertest');
+
+describe('Root test: ', function() {
+
+  describe('GET /', function(){
```
Comment (aronfiechter) [inline test/routes-test/root.js:21]: Space before `{`

**Review by aronfiechter — COMMENTED** [review]
We could add `_example/` to `.gitignore` before merging.

Most comments are about spacing, I was bored, sorry.

File: public/components/freelancer.js (hunk 7 above)
Comment (larabr) [reply public/components/freelancer.js:21]: I don't think so, I've always seen it like that, but never tried

File: README.md (hunk 1 above)
Comment (larabr) [reply README.md:1]: next sprint

File: models/Review.js (hunk 2 above)
Comment (larabr) [reply models/Review.js]: k

File: models/Review.js (hunk 3 above)
Comment (larabr) [reply models/Review.js:38]: the empty string is fine no?

File: models/Review.js (hunk 4 above)
Comment (larabr) [reply models/Review.js]: k

File: models/Tag.js (hunk 5 above)
Comment (larabr) [reply models/Tag.js]: k

File: package.json (hunk 6 above)
Comment (larabr) [reply package.json]: k

File: public/components/freelancer.js (hunk 7 above)
Comment (larabr) [reply public/components/freelancer.js:21]: it works but we put it nowhere and in html its nicer like that

**Comment by CrashingBrain** [comment]
I will add an hardcoded location ("Lugano || Bellinzona || Mendrisio") for the demo

File: public/components/homepage.js (hunk 8 above)
Comment (larabr) [reply public/components/homepage.js:94]: not there, check routing next sprint

File: public/css/style.css (hunk 9 above)
Comment (larabr) [reply public/css/style.css:1]: @jfida look into it for next sprint pls?

**Review by CrashingBrain — CHANGES_REQUESTED** [review]
Need to add hardcoded location in seed

File: public/js/ajax.js (hunk 10 above)
Comment (larabr) [reply public/js/ajax.js]: generalised with log of all status codes (in default), and handles specific cases of 200,201,205 only

File: routes/freelance/router.js (hunk 11 above)
Comment (larabr) [reply routes/freelance/router.js:32]: next sprint

**Review by CrashingBrain — APPROVED** [review]

File: seed_data/seedData.js (hunk 13 above)
Comment (larabr) [reply seed_data/seedData.js]: k to all
